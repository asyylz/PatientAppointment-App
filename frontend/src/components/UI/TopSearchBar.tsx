import { Link } from 'react-router-dom';
import classes from './TopSearchBar.module.css';
import { useSelector } from 'react-redux';

const TopSearchBar: React.FC = () => {
  const {
    entities: { token, data },
    status,
    error,
  } = useSelector((state: RootState) => state.currentUser);

  return (
    <>
      <div className={classes.topBar}>
        <div className={classes.search}>
          <input type="text" name="search" placeholder="search here" />
          <label htmlFor="search">
            <div className={classes.searchIcon}>
              <i className="fas fa-search"></i>
            </div>
          </label>
        </div>

        {token && data ? (
          <Link to="/logout" className={classes.user}>
            <i className="fas fa-bell"></i>
            <img src="./public/doctor.png" alt="" />
          </Link>
        ) : (
          <Link className={classes.login} to="/auth">
            Login
          </Link>
        )}
      </div>
    </>
  );
};

export default TopSearchBar;
