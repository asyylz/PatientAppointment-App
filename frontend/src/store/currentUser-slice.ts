import { createSlice, createAsyncThunk, createAction } from '@reduxjs/toolkit';
import axios, { AxiosError } from 'axios';
import { toastErrorNotify, toastSuccessNotify } from './../helper/ToastNotify';

const initialState: CurrentUser = {
  status: 'idle',
  token: '',
  userData: null,
};

/* ------------------------------------------------------ */
/*                        REGISTER                        */
/* ------------------------------------------------------ */
export const register = createAsyncThunk<
  CurrentUserPayload,
  FormData,
  { rejectValue: string }
>('signup', async (credentials, { rejectWithValue }) => {
  console.log(credentials);
  try {
    const response = await axios.post(
      'http://localhost:3000/api/v1/users/signup',
      credentials
    );
    toastSuccessNotify('Successfully registered!');
    console.log(response.data);
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      console.log(error.response?.data.message);
      toastErrorNotify(error.response?.data.message);
      if (axiosError.response?.status === 401) {
        toastErrorNotify(
          `${(axiosError.response.data as { message: string }).message}`
        );
      }
      return rejectWithValue(error.message);
    }
  }
});

/* ------------------------------------------------------ */
/*                          LOGIN                         */
/* ------------------------------------------------------ */
export const login = createAsyncThunk<
  CurrentUserPayload,
  { email: string; password: string },
  { rejectValue: string }
>('currentUser/login', async (credentials, { rejectWithValue }) => {
  console.log(credentials);
  try {
    const response = await axios.post(
      'http://localhost:3000/api/v1/users/login',
      credentials
    );
    toastSuccessNotify('Successfully login!');
    //console.log(response.data);
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      console.log(error);
      toastErrorNotify(error.response?.data.message);
      if (axiosError.response?.status === 401) {
        toastErrorNotify(
          `${(axiosError.response.data as { message: string }).message}`
        );
      }
      return rejectWithValue(error.message);
    }
  }
});

/* ------------------------------------------------------ */
/*                         LOGOUT                         */
/* ------------------------------------------------------ */
export const logout = createAsyncThunk<void, string, { rejectValue: string }>(
  'currentUser/logout',
  async (token: string, { rejectWithValue }) => {
    try {
      await axios.get('http://localhost:3000/api/v1/users/logout', {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      });
      toastSuccessNotify('Successfully logout!');
    } catch (error) {
      if (axios.isAxiosError(error)) {
        const axiosError = error as AxiosError;
        if (axiosError.response?.status === 401) {
          toastErrorNotify(
            `${(axiosError.response.data as { message: string }).message}`
          );
        }
        return rejectWithValue(error.message);
      }
    }
  }
);

/* ------------------------------------------------------ */
/*                         UPDATE                         */
/* ------------------------------------------------------ */
export const updateUserInfo = createAsyncThunk<
  CurrentUserPayload,
  { token: string; userUpdatedFormData: FormData },
  { rejectValue: string }
>('currentUser/updateProfile', async (tokenAndData, { rejectWithValue }) => {
  console.log(tokenAndData.userUpdatedFormData.get('address'));
  try {
    const response = await axios.patch(
      'http://localhost:3000/api/v1/users/updateUser',
      tokenAndData.userUpdatedFormData,
      {
        headers: {
          Authorization: `Bearer ${tokenAndData.token}`,
          'Content-Type': 'multipart/form-data',
        },
      }
    );

    toastSuccessNotify('Successfully updated!');
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      if (axiosError.response?.status === 401) {
        toastErrorNotify(
          `${(axiosError.response.data as { message: string }).message}`
        );
      }
      return rejectWithValue(error.message);
    }
    return rejectWithValue('An unexpected error occurred');
  }
});

/* ------------------------------------------------------ */
/*                         FORGET                         */
/* ------------------------------------------------------ */
export const forgotPassword = createAsyncThunk<
  CurrentUserPayload,
  { email: string },
  { rejectValue: string }
>('currentUser/forgotPassword', async (email, { rejectWithValue }) => {
  console.log(email);
  try {
    const response = await axios.post(
      'http://localhost:3000/api/v1/users/forgotPassword',
      email
    );
    toastSuccessNotify(`Email sent to ${email.email} successfully!`);
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      const axiosError = error as AxiosError;
      if (axiosError.response?.status === 401) {
        toastErrorNotify(
          `${(axiosError.response.data as { message: string }).message}`
        );
      }
      return rejectWithValue(error.message);
    }
    return rejectWithValue('An unexpected error occurred');
  }
});

/* ------------------------------------------------------ */
/*                          RESET                         */
/* ------------------------------------------------------ */
export const resetPassword = createAsyncThunk<
  CurrentUserPayload,
  { password: string; passwordConfirm: string; resetToken: string },
  { rejectValue: string }
>(
  'currentUser/resetPassword',
  async ({ password, passwordConfirm, resetToken }, { rejectWithValue }) => {
    console.log(password, passwordConfirm, resetToken);
    console.log('Asiye');

    try {
      const response = await axios.patch(
        `http://localhost:3000/api/v1/users/resetPassword/${resetToken}`,
        {
          password,
          passwordConfirm,
        }
      );
      toastSuccessNotify(`Your password successfully re-set`);
      console.log(response.data);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        if (error.response?.status === 400) {
          toastErrorNotify(
            `${(error.response.data as { message: string }).message}`
          );

          return rejectWithValue(error.response?.data?.message);
        }
      }
      return rejectWithValue('An unexpected error occurred');
    }
  }
);
/* ------------------------------------------------------ */
/*                     UPDATE PASSWORD                    */
/* ------------------------------------------------------ */
export const updatePassword = createAsyncThunk<
  CurrentUserPayload,
  UpdatedUserPasswordAndToken,
  { rejectValue: string }
>('currentUser/updatePassword', async (data, { rejectWithValue }) => {
  const { oldPassword, newPassword, confirmNewPassword, token } = data;
  console.log(oldPassword, newPassword, confirmNewPassword, token);

  try {
    const response = await axios.patch(
      'http://localhost:3000/api/v1/users/updateMyPassword',
      { oldPassword, newPassword, confirmNewPassword },
      {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      }
    );
    toastSuccessNotify(`Your password successfully updated`);
    console.log(response.data);
    return response.data;
  } catch (error) {
    if (axios.isAxiosError(error)) {
      if (error.response?.status === 400) {
        toastErrorNotify(
          `${(error.response.data as { message: string }).message}`
        );

        return rejectWithValue(error.response?.data?.message);
      }
    }
    return rejectWithValue('An unexpected error occurred');
  }
});

/* ------------------------------------------------------ */
/*                          SLICE                         */
/* ------------------------------------------------------ */
const currentUserSlice = createSlice({
  name: 'currentUser',
  initialState,
  reducers: {
    // setImagePath: (state, action: PayloadAction<string>) => {
    //   state.image = action.payload;
    // },
  },
  extraReducers: (builder) => {
    builder
      .addCase(login.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(login.fulfilled, (state, action) => {
        const { token, data } = action.payload;
        state.status = 'success';
        state.token = token;
        state.userData = data.user;
        state.error = null;
      })
      .addCase(login.rejected, (state, action) => {
        state.status = 'failed';
        console.log(action.error.message);
        state.error = action.error.message || 'Login failed';
      })
      .addCase(register.fulfilled, (state, action) => {
        state.status = 'success';
        state.token = action.payload.token;
        state.userData = action.payload.data.user;
        state.error = null;
      })
      .addCase(resetPassword.fulfilled, (state, action) => {
        state.status = 'success';
        state.token = action.payload.token;
        state.userData = action.payload.data.user;
        state.error = null;
      })
      .addCase(updatePassword.fulfilled, (state, action) => {
        state.status = 'success';
        state.token = action.payload.token;
        state.userData = action.payload.data.user;
        state.error = null;
      })
      .addCase(logout.fulfilled, (state) => {
        state.status = 'logout success';
        state.token = '';
        state.userData = null;
        state.error = null;
      })
      .addCase(logout.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload || 'Logout failed';
      })
      .addCase(updateUserInfo.fulfilled, (state, action) => {
        state.status = 'update success';
        console.log(action.payload);
        //state.userData = action.payload.data.user;
        state.userData = action.payload.data.user;
        state.error = null;
      })
      .addCase(updateUserInfo.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload || 'Update failed';
      })
      .addCase('currentUser/stateToIdle', (state) => {
        state.status = 'idle'; // Reset status to 'idle' on successful logout
      });
  },
});

// Export actions and reducer
export const logoutSuccess = createAction('currentUser/stateToIdle');

//export const { setImagePath } = currentUserSlice.actions;
export default currentUserSlice.reducer;
