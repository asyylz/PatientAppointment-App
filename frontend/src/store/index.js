import { configureStore } from '@reduxjs/toolkit';

const store = configureStore({
  reducer: { departments: departmentsSlice.reducer }, // root reducer
});

export default store;
