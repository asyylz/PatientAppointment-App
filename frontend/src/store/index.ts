import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import departmentsReducer from './departments-slice';
import doctorsReducer from './doctors-slice';
import reviewsReducer from './reviews-slice';
import currentUserReducer from './currentUser-slice';
//import { useDispatch } from 'react-redux';
import searchReducer from './search-slice';
import appointmentsForDoctorReducer from './appointmentsForDoctor-slice';
import appointmentsForPatientReducer from './appointmentsForPatient-slice';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web

// const rootPersistConfig = {
//   key: 'root',
//   storage,
// };

// const persistedReducer = persistReducer(rootPersistConfig, currentUserReducer);

// const store = configureStore({
//   reducer: {
//     currentUser: persistedReducer,
//     departments: departmentsReducer,
//     doctors: doctorsReducer,
//     reviews: reviewsReducer,
//     search: searchReducer,
//     appointmentsForDoctor: appointmentsForDoctorReducer,
//     appointmentsForPatient: appointmentsForPatientReducer,
//   },
//   middleware: (getDefaultMiddleware) =>
//     getDefaultMiddleware({
//       serializableCheck: {
//         ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
//         ignoredActionPaths: ['meta.arg', 'payload.timestamp'],
//         ignoredPaths: ['items.dates'],
//       },
//     }),
//   devTools: process.env.NODE_ENV !== 'production',
// });

// const rootPersistConfig = {
//   key: 'root',
//   storage,
//   whitelist: ['currentUser', 'doctors'], // Only persist the 'currentUserSlice' slice
// };

const authPersistConfig = {
  key: 'currentUser',
  storage: storage,
  // blacklist: ['status', 'error'],
};
const doctorPersistConfig = {
  key: 'doctors',
  storage: storage,
  // blacklist: ['status','error']
};
export const allReducers = combineReducers({
  currentUser: persistReducer(authPersistConfig, currentUserReducer),
  departments: departmentsReducer,
  doctors: persistReducer(doctorPersistConfig, doctorsReducer),
  reviews: reviewsReducer,
  search: searchReducer,
  appointmentsForDoctor: appointmentsForDoctorReducer,
  appointmentsForPatient: appointmentsForPatientReducer,
});

//const persistedReducer = persistReducer(rootPersistConfig, allReducers);

const store = configureStore({
  reducer: allReducers,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
});

export type AppDispatch = typeof store.dispatch;
//export const useAppDispatch = useDispatch.withTypes<AppDispatch>();
export type AppStore = typeof store;
export type RootState = ReturnType<AppStore['getState']>;

export const persistor = persistStore(store);

export default store;
