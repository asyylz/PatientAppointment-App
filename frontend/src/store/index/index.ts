import { configureStore, combineReducers } from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import departmentsReducer from '../departments-slice/departments-slice';
import doctorsReducer from '../doctors-slice/doctors-slice';
import reviewsReducer from './../reviews-slice/reviews-slice';
import currentUserReducer from '../currentUser-slice/currentUser-slice';
import searchReducer from '../search-slice/search-slice';
import appointmentsForDoctorReducer from '../appointmentsForDoctor-slice/appointmentsForDoctor-slice';
import appointmentsForPatientReducer from '../appointmentsForPatient-slice/appointmentsForPatient-slice';
import storage from 'redux-persist/lib/storage'; // defaults to localStorage for web

const authPersistConfig = {
  key: 'currentUser',
  storage: storage,
  // blacklist: ['status', 'error'],
};

const doctorPersistConfig = {
  key: 'doctors',
  storage: storage,
  // blacklist: ['status','error']
};

export const allReducers = combineReducers({
  currentUser: persistReducer(authPersistConfig, currentUserReducer),
  departments: departmentsReducer,
  doctors: persistReducer(doctorPersistConfig, doctorsReducer),
  reviews: reviewsReducer,
  search: searchReducer,
  appointmentsForDoctor: appointmentsForDoctorReducer,
  appointmentsForPatient: appointmentsForPatientReducer,
});

const store = configureStore({
  reducer: allReducers,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [
          FLUSH,
          REHYDRATE,
          PAUSE,
          PERSIST,
          PURGE,
          REGISTER,
          'appointmentsForPatient/fetchWithId/rejected',
          'appointmentsForDoctor/fetchWithId/rejected',
          'departments/fetch/rejected',
          'doctors/fetch/rejected',
        ],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
});

export function setupStore(reducer?: any, preloadedState?: Partial<RootState>) {
  //console.log('from index store', preloadedState); // i can see my preloaded state here  with mock data
  const store = configureStore({
    reducer: reducer,
    preloadedState,
  });

  //console.log(store.getState()); // but still cant see current user but can see the rest of the states
  return store;
}

export type AppDispatch = typeof store.dispatch;
export type AppStore = typeof store;
export type RootState = ReturnType<AppStore['getState']>;

export const persistor = persistStore(store);

export default store;
