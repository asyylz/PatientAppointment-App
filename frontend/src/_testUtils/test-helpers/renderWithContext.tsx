import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../../store/index';
import type { AppStore, RootState } from '../../store/index';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  reducer?: any;
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const { preloadedState, reducer } = extendedRenderOptions;
  // Ensure that the store is being correctly initialized with preloadedState
  const store = setupStore(reducer, preloadedState);
  //console.log(store.getState()); // Verify the state of the store

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>{children}</Provider>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...extendedRenderOptions }),
  };
}
