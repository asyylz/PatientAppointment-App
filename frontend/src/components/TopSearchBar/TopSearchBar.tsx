import classes from './TopSearchBar.module.css';
import { useSelector, useDispatch } from 'react-redux';
import { AppDispatch } from '../../store/index';
import { useEffect, useState } from 'react';
import { setSearch } from '../../store/search-slice/search-slice';
import { performLogout } from '../../store/currentUser-slice/currentUser-slice';

const TopSearchBar: React.FC = () => {
  const dispatch: AppDispatch = useDispatch();
  const [searchInput, setSearchInput] = useState('');

  const { token, userData } = useSelector(
    (state: RootState) => state.currentUser
  );

  useEffect(() => {
    const timeout = setTimeout(() => dispatch(setSearch(searchInput)), 500);
    return () => {
      clearTimeout(timeout);
    };
  }, [dispatch, searchInput]);

  const handleLogout = async () => {
    await dispatch(performLogout());
  };

  return (
    <>
      <div className={classes['top__bar']}>
        <div className={classes['top__bar--search']}>
          <input
            className={classes['top__bar--search-input']}
            type="text"
            name="search"
            placeholder="Search here"
            onChange={(e) => setSearchInput(e.target.value)}
          />

          <div className={classes['top__bar--search-icon']}>
            <i className="fas fa-search"></i>
          </div>
        </div>

        {token && userData && (
          <div className={classes['top__bar--search-user']}>
            <h5>
              {userData?.role === 'doctor'
                ? `Dr. ${userData?.name}`
                : userData?.name}
            </h5>
            <i className="fas fa-bell"></i>
            <img src={userData.image} alt="User" />
            <button onClick={handleLogout}>Logout</button>
          </div>
          // ) : (
          //   <button onClick={() => navigate('/auth')}>Login</button>
        )}
      </div>
    </>
  );
};

export default TopSearchBar;
