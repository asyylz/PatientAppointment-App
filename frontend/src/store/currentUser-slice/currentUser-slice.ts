import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import { toastSuccessNotify } from '../../helper/ToastNotify';
import {
  axiosInterceptorsWithToken,
  axiosInterceptorsWithoutToken,
} from '../../services/axiosInterceptors';
import {
  startTokenCheckInterval,
  stopTokenCheckInterval,
} from '../../helper/timers';

import { AppDispatch } from '../index';

const initialState: CurrentUser = {
  status: 'idle',
  token: '',
  userData: null,
};

/* ------------------------------------------------------ */
/*                        REGISTER                        */
/* ------------------------------------------------------ */
export const register = createAsyncThunk<CurrentUserPayload, Credentials>(
  'currentUser/register',
  async (credentials) => {

    const response = await axiosInterceptorsWithoutToken.post(
      'http://localhost:3000/api/v1/users/signup',
      credentials
    );
    toastSuccessNotify('Successfully registered!');
    return response.data;
  }
);

/* ------------------------------------------------------ */
/*                          LOGIN                         */
/* ------------------------------------------------------ */
export const login = createAsyncThunk<
  CurrentUserPayload,
  { email: string; password: string }
>('currentUser/login', async (credentials) => {
  const response = await axiosInterceptorsWithoutToken.post(
    'http://localhost:3000/api/v1/users/login',
    credentials,
    { withCredentials: true }
  );

  startTokenCheckInterval();
  toastSuccessNotify('Successfully login!');
  return response.data;
});

/* ------------------------------------------------------ */
/*                         LOGOUT                         */
/* ------------------------------------------------------ */
export const performLogout = () => async (dispatch: AppDispatch) => {
  const response = await axiosInterceptorsWithToken.get(
    'http://localhost:3000/api/v1/users/logout'
  );
  if (response.data.status === 'success') {
    console.log('asiye');
    dispatch(logout());
    stopTokenCheckInterval();
  }

  return response.data;
};

/* ------------------------------------------------------ */
/*                         UPDATE                         */
/* ------------------------------------------------------ */
export const updateUserInfo = createAsyncThunk<
  CurrentUserPayload,
  FormData,
  { rejectValue: string }
>('currentUser/updateProfile', async (userUpdatedFormData) => {
  //try {
  const response = await axiosInterceptorsWithToken.patch(
    'http://localhost:3000/api/v1/users/updateUser',
    userUpdatedFormData
  );

  toastSuccessNotify('Your profile successfully updated!');
  return response.data;
});

/* ------------------------------------------------------ */
/*                      REFRESH TOKEN                     */
/* ------------------------------------------------------ */
export const refreshSession = createAsyncThunk<
  CurrentUserPayload,
  void,
  { rejectValue: string }
>('currentUser/refreshSession', async () => {
  const response = await axiosInterceptorsWithToken.post(
    'http://localhost:3000/api/v1/users/refresh-session'
  );

  toastSuccessNotify('Your session has been extended another 15 mins!');
  return response.data;
});

/* ------------------------------------------------------ */
/*                          SLICE                         */
/* ------------------------------------------------------ */
const currentUserSlice = createSlice({
  name: 'currentUser',
  initialState,
  reducers: {
    logout(state) {
      state.status = 'idle';
      state.token = '';
      state.userData = null;
      state.error = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(login.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(login.fulfilled, (state, action) => {
        const { token, data } = action.payload;
        state.status = 'login success';
        state.token = token;
        state.userData = data.user;
        state.error = null;
      })
      .addCase(login.rejected, (state, action) => {
        state.status = 'login failed';
        state.error = action.error?.message || 'Login failed!';
      })
      .addCase(refreshSession.fulfilled, (state, action) => {
        const { token, data } = action.payload;
        state.status = 'refresh success';
        state.token = token;
        console.log(token);
        state.userData = data.user;
        state.error = null;
      })
      .addCase(refreshSession.rejected, (state, action) => {
        state.status = 'refresh failed';
        state.error = action.error.message || 'Refresh failed';
        state.token = ''; // need to be checked
      })
      .addCase(register.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(register.fulfilled, (state, action) => {
        state.status = 'register success';
        state.token = action.payload.token;
        state.userData = action.payload.data.user;
        state.error = null;
      })
      .addCase(register.rejected, (state, action) => {
        state.status = 'register failed';
        state.error = action.error.message || 'Register failed';
      })
      .addCase(updateUserInfo.fulfilled, (state, action) => {
        state.status = 'update success';
        state.userData = action.payload.data.user;
        state.error = null;
      })
      .addCase(updateUserInfo.rejected, (state, action) => {
        state.status = 'update failed';
        state.error = action.error?.message || 'Updating user info failed';
      })
      .addCase('currentUser/stateToIdle', (state) => {
        state.status = 'idle'; // Reset status to 'idle' on successful logout
      });
  },
});

// Export actions and reducer
//export const logoutSuccess = createAction('currentUser/stateToIdle');

export const { logout } = currentUserSlice.actions;

export default currentUserSlice.reducer;
