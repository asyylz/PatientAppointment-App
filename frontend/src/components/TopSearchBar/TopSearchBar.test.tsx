import '@testing-library/jest-dom';
import { act, fireEvent, screen, waitFor } from '@testing-library/react';
import { combineReducers } from 'redux';
import TopSearchBar from '../TopSearchBar/TopSearchBar.tsx';
import searchReducer from '../../store/search-slice/search-slice';
import currentUserReducer from '../../store/currentUser-slice/currentUser-slice';
//import * as router from 'react-router-dom';
import * as currentUserSlice from '../../store/currentUser-slice/currentUser-slice';
import { renderWithProviders } from '../../_testUtils/test-helpers/renderWithContext';
import { PersistPartial } from 'redux-persist/es/persistReducer';
import { AppStore } from '../../store/index';


jest.mock('../../helper/ToastNotify', () => ({
  toastSuccessNotify: jest.fn(),
}));
jest.mock('redux-persist', () => {
  const actual = jest.requireActual('redux-persist');
  return {
    ...actual,
    persistReducer: jest.fn().mockImplementation((_config, reducer) => reducer),
  };
});

// Mock useNavigate
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));
// Start store with initial state for `currentUser`
const initialCurrentUserState: CurrentUser = {
  status: 'login success',
  token:
    'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjY2NzM2NjJmYmQ0MmE5NjZiNzVkZWM5MiIsImlhdCI6MTcyMzc1NDAzOSwiZXhwIjoxNzIzNzU1ODM5fQ.4Nmyge8591WiGN3wEijyn-9Tx7A0GCBu5w2a0CgVaQs',
  userData: {
    _id: '6673662fbd42a966b75dec92',
    name: 'Asiye',
    email: 'alice@test.com',
    role: 'patient',
    __v: 0,
    DOB: '1986-01-22T00:00:00.000Z',
    image:
      'https://patient-appointment-system.s3.eu-west-2.amazonaws.com/1722605356602-asiye%20serife.jpeg',
    policy: true,
  },
  error: null,
};

const reducers = combineReducers({
  currentUser: currentUserReducer,
  search: searchReducer,
  // other reducers...
});

describe('TopSearchBar Component', () => {
 // const navigate = jest.fn();
  let store: AppStore;
  let asFragment: () => DocumentFragment;

  beforeEach(() => {
    //jest.spyOn(router, 'useNavigate').mockImplementation(() => navigate);
    const result = renderWithProviders(<TopSearchBar />, {
      reducer: reducers,
      preloadedState: { currentUser: initialCurrentUserState } as Partial<{
        currentUser: CurrentUser & PersistPartial;
      }>,
    });

    store = result.store;
    asFragment = result.asFragment;
    jest.spyOn(store, 'dispatch');
  });

  afterEach(() => {
    jest.clearAllMocks();
    //jest.restoreAllMocks();
  });
  /* -------------------------- - ------------------------- */
  it('1--Matches the DOM snapshot', () => {
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Asiye')).toBeInTheDocument();
    expect(screen.getByText('Logout')).toBeInTheDocument();
  });
  /* -------------------------- - ------------------------- */
  it('2--Search input typed and trigger departments to be filtered', async () => {
    const searchInput = screen.getByPlaceholderText('Search here');
    fireEvent.change(searchInput, { target: { value: 'Anesthesiology' } });

    expect(store.dispatch).not.toHaveBeenCalledWith({
      type: 'search/setSearch',
      payload: 'Anesthesiology',
    }); // time for debounce, so  no call yet

    // Here we expilicitly update state
    await act(async () => {
      store.dispatch({
        type: 'search/setSearch',
        payload: 'Anesthesiology',
      });
    });
    await waitFor(() => {
      expect(store.dispatch).toHaveBeenCalledWith({
        type: 'search/setSearch',
        payload: 'Anesthesiology',
      });
    });
    expect(store.getState().search).toEqual('Anesthesiology');
  });
  /* -------------------------- - ------------------------- */
  it('3--Logout button should navigate user to home page', async () => {
    // Mock the performLogout action
    const mockPerformLogout = jest
      .fn()
      .mockResolvedValue({ status: 'success' });
    jest
      .spyOn(currentUserSlice, 'performLogout')
      .mockImplementation(() => () => mockPerformLogout());

    // Find and click the logout button
    const logoutButton = screen.getByText('Logout');
    fireEvent.click(logoutButton);

    // Wait for performLogout to be called
    await waitFor(() => {
      expect(mockPerformLogout).toHaveBeenCalled();
    });
  });
});
