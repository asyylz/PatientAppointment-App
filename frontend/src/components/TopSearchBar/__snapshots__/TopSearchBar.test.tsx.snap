// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopSearchBar Component 1--Matches the DOM snapshot 1`] = `
<DocumentFragment>
  <div
    class="top__bar"
  >
    <div
      class="top__bar--search"
    >
      <input
        class="top__bar--search-input"
        name="search"
        placeholder="Search here"
        type="text"
      />
      <div
        class="top__bar--search-icon"
      >
        <i
          class="fas fa-search"
        />
      </div>
    </div>
    <div
      class="top__bar--search-user"
    >
      <h5>
        Asiye
      </h5>
      <i
        class="fas fa-bell"
      />
      <img
        alt="User"
        src="https://patient-appointment-system.s3.eu-west-2.amazonaws.com/1722605356602-asiye%20serife.jpeg"
      />
      <button>
        Logout
      </button>
    </div>
  </div>
</DocumentFragment>
`;
